lang/py/avro/compatibility.py

```diff
@@ -0,0 +1,318 @@
+        return SchemaCompatibilityResult(
+            compatibility=compat, incompatibilities=merged, messages=messages, locations=locations
+        )
+
+    @staticmethod
+    def compatible():
+        return SchemaCompatibilityResult(SchemaCompatibilityType.compatible)
+
+    @staticmethod
+    def incompatible(incompat_type: SchemaIncompatibilityType, message: str, location: List[str]):
+        locations = "/".join(location)
+        if len(location) > 1:
+            locations = locations[1:]
+        ret = SchemaCompatibilityResult(
+            compatibility=SchemaCompatibilityType.incompatible,
+            incompatibilities=[incompat_type],
+            locations={locations},
+            messages={message},
+        )
+        return ret
+
+    def __str__(self):
+        return f"{self.compatibility}: {self.messages}"
+
+
+class ReaderWriter:
+    def __init__(self, reader: Schema, writer: Schema) -> None:
+        self.reader, self.writer = reader, writer
+
+    def __hash__(self) -> SchemaType.INT:
+        return id(self.reader) ^ id(self.writer)
+
+    def __eq__(self, other) -> bool:
+        if not isinstance(other, ReaderWriter):
+            return False
+        return self.reader is other.reader and self.writer is other.writer
+
+
+class ReaderWriterCompatibilityChecker:
+    ROOT_REFERENCE_TOKEN = "/"
+
+    def __init__(self):
+        self.memoize_map: Dict[ReaderWriter, SchemaCompatibilityResult] = {}
+
+    def get_compatibility(
+        self,
+        reader: Schema,
+        writer: Schema,
+        reference_token: str = ROOT_REFERENCE_TOKEN,
+        location: Optional[List[str]] = None
+    ) -> SchemaCompatibilityResult:
+        if location is None:
+            location = []
+        location.append(reference_token)
+        pair = ReaderWriter(reader, writer)
+        if pair in self.memoize_map:
+            result = self.memoize_map[pair]
+            if result.compatibility is SchemaCompatibilityType.recursion_in_progress:
+                result = SchemaCompatibilityResult.compatible()
+        else:
+            self.memoize_map[pair] = SchemaCompatibilityResult()
+            result = self.calculate_compatibility(reader, writer, location)
+            self.memoize_map[pair] = result
+        location.pop()
+        return result
+
+    # pylSchemaType.INT: disable=too-many-return-statements
+    def calculate_compatibility(
+        self,
+        reader: Schema,
+        writer: Schema,
+        location: List[str],
+    ) -> SchemaCompatibilityResult:
+        assert reader is not None
+        assert writer is not None
+        result = SchemaCompatibilityResult.compatible()
+        if reader.type == writer.type:
+            if reader.type in {
+                SchemaType.NULL, SchemaType.BOOLEAN, SchemaType.INT,
+                SchemaType.LONG, SchemaType.FLOAT, SchemaType.DOUBLE,
+                SchemaType.BYTES, SchemaType.STRING
+            }:
```

Comment:
Should probably just have a module constant for the set of primitive types.

Reply:
Added as module constant